### ISSUE TITLE
Dependency Parser Training Error

### ISSUE THREAD

**Issue by kevinrosenberg21** [explosion/spaCy#2052]
## Info about spaCy

* **spaCy version:** 2.0.5
* **Models:** en, es
* **Platform:** Linux-4.13.0-36-generic-x86_64-with-debian-stretch-sid
* **Python version:** 3.5.2

Hi,

I'm training a dependency parser on my custom semantics. During training I get the following error: 

> Could not find a gold-standard action to supervise thedependency parser. The GoldParse was projective. The transition system has 207 actions.

Before that I was getting an error saying that the tree was non-projective and had to change my training code to:

```
gold_data = [GoldParse(doc, heads=annotations['heads'], deps=annotations['deps'], make_projective = True)]
nlp_parser.update([text], gold_data, sgd=optimizer)
```
I'm using the dependency parser to find relationships between the entities. The entire document is a single sentence, and can be quite large. Specifically, I'm training a NER model on the one hand to recognize all the entities in a resume, and a dependency parser to find out how those entities relate to one another.
 
My question is: is this an error that shows up because of an error in the data/program or could it be that the problem is just not suitable for dependency parsing and need to look for another way to train a model to find relationships between the entities?

Thank you

**Comment by honnibal** [comment]
tl;dr: Can you give a sample of the text with the heads and dependencies you’ve assigned as the gold standard?

From a theoretical perspective, your approach does make sense. However the implementation of the parser is designed towards syntactic relationships, which are normally between words fairly close together. So, the features in the parser might not work well for your task, and it might be unexpectedly slow.

The other thing about the parsing algorithm is that it’s fairly complicated. If you haven’t seen it yet, I would suggest you have a look at this blog post: https://explosion.ai/blog/parsing-english-in-python . The post is old, but the parser still uses the same transition-based approach supervised by a “dynamic oracle”. It’s just that we use a neural network to optimise, instead of this simpler ML algorithm.

These two sections are especially relevant:

Greedy transition-based parsing: https://explosion.ai/blog/parsing-english-in-python#greedy-transition-based-parsing

Training: https://explosion.ai/blog/parsing-english-in-python#training

The gist of this is that we’re setting up an initial state that has a stack, a queue of words, and a set of dependency arcs. We then define some fixed set of actions that we’ll use to transition from one state to another. This lets us map the parsing task to the task of predicting a set of actions that end in with a desirable parse tree.

The important thing to understand for your error is that the training algorithm requires us to assign a “cost” to each potential action, where the cost is the number of additional errors that action would introduce. In other words: What’s the score of the best parse we can make from this state? Okay, what’s the score of the best parse we can make if we apply this action to this state? The cost of the action is the difference between the two.

So, your error is saying that given the state you’re in and the gold-parse you’ve assigned to the sentence, none of the actions result in zero cost. This might occur if there’s no way to derive the gold-standard you assigned, given the actions the parser has (for instance if it doesn’t have a label in your gold-standard).

Another thing to keep in mind is that the parser has to build a tree that covers every token. It’s possible to underspecify the tree, in which case the parser will have no guidance for some of the arcs. In your case, if you’re trying to only extract relations between some entities in a whole document, you might end up with the vast majority of tokens underspecified. This will probably be very hard for the parser to learn from.

**Comment by rbhambriiit** [comment]
I am getting similar error while training an NER system.

`ipython-input-3-7bd9ad53e8a8> in train_ner(model, n_iter)
     51                     drop=0.5,  # dropout - make it harder to memorise data
     52                     sgd=optimizer,  # callable to update weights
---> 53                     losses=losses)
     54             print(losses)
     55 

/home/703204188/venv/my_project/local/lib/python2.7/site-packages/spacy/language.pyc in update(self, docs, golds, drop, sgd, losses)
    413                 continue
    414             grads = {}
--> 415             proc.update(docs, golds, drop=drop, sgd=get_grads, losses=losses)
    416             for key, (W, dW) in grads.items():
    417                 sgd(W, dW, key=key)

nn_parser.pyx in spacy.syntax.nn_parser.Parser.update()

nn_parser.pyx in spacy.syntax.nn_parser.Parser._init_gold_batch()

transition_system.pyx in spacy.syntax.transition_system.TransitionSystem.get_oracle_sequence()

transition_system.pyx in spacy.syntax.transition_system.TransitionSystem.set_costs()

ValueError: Could not find a gold-standard action to supervise the entity recognizer. The transition system has 53 actions.`


Right now I have a large doc with multiple sentences. While training I have specified the full large doc - will it help if I do this at a sentence level.

What is the recommendation for NER systems. Imagine a magazine cover of 20 pages with different entities of let's say - sports persons. Should I keep each sentence with the matched entities as a doc and train? I started with feeding the entire 20 pages as the doc, but dont mind learning at lower level - say sentences - if that helps.

@honnibal  - Please guide.

**Comment by rbhambriiit** [comment]
I did receive the messages similar to https://support.prodi.gy/t/cryptic-error-message-when-training-a-model-after-evaluation/263

Unfortunately, I can not share the exact debug logs. But they look like:

 print all tokens in the doc as list - followed by 
list of lots of BILUO's - followed by
list of numbers such as [4, 5, 2] ( i guess the transition index) - followed by
Self labels.

**Comment by rbhambriiit** [comment]
As per the above link on Prodigy - I think this statement touches the real issue

"""
I’m fairly sure the issue you’re seeing would come down to training from a Doc object that already has the NER annotations set.
"""

**Comment by rbhambriiit** [comment]
Update: 
Now when I think about it, it makes sense it give raw text and not already parsed doc to a blank NER model.

I can confirm that replacing doc with raw text , fixes the case for me.

```
            for text, annotations in TRAIN_DATA:
#                 print type(text), len(text), annotations
                nlp.update(
                    [unicode(text)],  # batch of texts
                    [annotations],  # batch of annotations
                    drop=0.5,  # dropout - make it harder to memorise data
                    sgd=optimizer,  # callable to update weights
                    losses=losses)
            print(losses)

```

(earlier instead of  [unicode(text)], i had a [doc])
